x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/AESKeyUtils.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.repositories.encrypted;
+
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Base64;
+
+public final class AESKeyUtils {
+    public static final int KEY_LENGTH_IN_BYTES = 32; // 256-bit AES key
+    public static final int WRAPPED_KEY_LENGTH_IN_BYTES = KEY_LENGTH_IN_BYTES + 8; // https://www.ietf.org/rfc/rfc3394.txt section 2.2
+    // parameter for the KDF function, it's a funny and unusual iter count larger than 60k
+    private static final int KDF_ITER = 61616;
+    // the KDF algorithm that generate the symmetric key given the password
+    private static final String KDF_ALGO = "PBKDF2WithHmacSHA512";
+    // The Id of any AES SecretKey is the AES-Wrap-ciphertext of this fixed 32 byte wide array.
+    // Key wrapping encryption is deterministic (same plaintext generates the same ciphertext)
+    // and the probability that two different keys map the same plaintext to the same ciphertext is very small
+    // (2^-256, much lower than the UUID collision of 2^-128), assuming AES is indistinguishable from a pseudorandom permutation.
+    private static final byte[] KEY_ID_PLAINTEXT = "wrapping known text forms key id".getBytes(StandardCharsets.UTF_8);
+
+    public static byte[] wrap(SecretKey wrappingKey, SecretKey keyToWrap) throws NoSuchPaddingException,
+        NoSuchAlgorithmException,
+        InvalidKeyException,
+        IllegalBlockSizeException {
+        if (false == "AES".equals(wrappingKey.getAlgorithm())) {
+            throw new IllegalArgumentException("wrappingKey argument is not an AES Key");
+        }
+        if (false == "AES".equals(keyToWrap.getAlgorithm())) {
+            throw new IllegalArgumentException("toWrapKey argument is not an AES Key");
+        }
+        Cipher c = Cipher.getInstance("AESWrap");
+        c.init(Cipher.WRAP_MODE, wrappingKey);
+        return c.wrap(keyToWrap);
+    }
+
+    public static SecretKey unwrap(SecretKey wrappingKey, byte[] keyToUnwrap) throws NoSuchPaddingException,
+        NoSuchAlgorithmException,
+        InvalidKeyException {
+        if (false == "AES".equals(wrappingKey.getAlgorithm())) {
```

Comment:
It would probably fail, these checks serve more as documentation than anything else.
I have a slight preference to keep them in, they're not too distracting, I think.

Reply:
>  these checks serve more as documentation than anything else.

+1 to having them as documentation, but maybe turn them into `assert`s then to make it clear that this is documentation? (just a suggestion, I'm fine with it either way)

Reply:
I looked into it a bit more.
Indeed **the wrapping** key must be `AES` (or `Rijndael`), otherwise the Sun security provider throws `InvalidKeyException`. **The wrapped** key algorithm does not really matter as long as the byte array of the encoded key conforms to the length requirements of the AESWrap algorithm.
Given this, I've transformed all the `IllegalArgumentExceptions` into asserts because asserts better reflect the nuance that it's more an expectation given the calling context rather than a strong argument validation.